php-checks/src/main/java/org/sonar/php/checks/LoopExecutingAtMostOnceCheck.java

```diff
@@ -0,0 +1,151 @@
+import com.google.common.collect.ArrayListMultimap;
+import com.google.common.collect.Iterables;
+import com.google.common.collect.ListMultimap;
+import com.google.common.collect.Sets;
+import java.util.Set;
+import org.sonar.check.Rule;
+import org.sonar.php.cfg.CfgBlock;
+import org.sonar.php.cfg.CfgBranchingBlock;
+import org.sonar.php.cfg.ControlFlowGraph;
+import org.sonar.php.tree.impl.PHPTree;
+import org.sonar.plugins.php.api.tree.CompilationUnitTree;
+import org.sonar.plugins.php.api.tree.Tree;
+import org.sonar.plugins.php.api.tree.statement.BreakStatementTree;
+import org.sonar.plugins.php.api.tree.statement.ForEachStatementTree;
+import org.sonar.plugins.php.api.tree.statement.ForStatementTree;
+import org.sonar.plugins.php.api.tree.statement.GotoStatementTree;
+import org.sonar.plugins.php.api.tree.statement.ReturnStatementTree;
+import org.sonar.plugins.php.api.tree.statement.ThrowStatementTree;
+import org.sonar.plugins.php.api.visitors.PHPVisitorCheck;
+import org.sonar.plugins.php.api.visitors.PreciseIssue;
+
+import static org.sonar.php.checks.utils.SyntacticEquivalence.areSyntacticallyEquivalent;
+import static org.sonar.php.tree.TreeUtils.findAncestorWithKind;
+import static org.sonar.php.tree.TreeUtils.isDescendant;
+
+@Rule(key = LoopExecutingAtMostOnceCheck.KEY)
+public class LoopExecutingAtMostOnceCheck extends PHPVisitorCheck {
+
+  public static final String KEY = "S1751";
+  private static final String MESSAGE = "Refactor this loop to do more than one iteration.";
+  private static final Set<Tree.Kind> LOOPS = Sets.immutableEnumSet(
+    Tree.Kind.WHILE_STATEMENT,
+    Tree.Kind.DO_WHILE_STATEMENT,
+    Tree.Kind.FOR_STATEMENT,
+    Tree.Kind.FOREACH_STATEMENT);
+
+  private ListMultimap<Tree, Tree> reportedLoops = ArrayListMultimap.create();
+
+  @Override
+  public void visitCompilationUnit(CompilationUnitTree tree) {
+    reportedLoops.clear();
+    super.visitCompilationUnit(tree);
+    reportIssues();
+  }
+
+  private void reportIssues() {
+    reportedLoops.asMap().forEach((loop, jumps) -> {
+      PreciseIssue preciseIssue = context().newIssue(this, ((PHPTree) loop).getFirstToken(), MESSAGE);
+      jumps.forEach(jump -> preciseIssue.secondary(jump, "loop exit"));
+    });
+  }
+
+  @Override
+  public void visitBreakStatement(BreakStatementTree tree) {
+    checkJump(tree);
+  }
+
+  @Override
+  public void visitReturnStatement(ReturnStatementTree tree) {
+    checkJump(tree);
+  }
+
+  @Override
+  public void visitThrowStatement(ThrowStatementTree tree) {
+    checkJump(tree);
+  }
+
+  @Override
+  public void visitGotoStatement(GotoStatementTree tree) {
+    checkJump(tree);
+  }
+
+  private void checkJump(Tree tree) {
+    Tree loop = findAncestorWithKind(tree, LOOPS);
+    if (loop != null && !canExecuteMoreThanOnce(loop) && !isForEachReturningFirstElement(loop, tree)) {
+      reportedLoops.put(loop, tree);
+    }
+  }
+
+  private static boolean isForEachReturningFirstElement(Tree loop, Tree jump) {
```

Comment:
Should we also check that this return is the only statement of the loop?

Reply:
ok